{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from utils.utils import read_json_file, save_json_file\n",
    "from llms.embedding_model import EmbeddingModel\n",
    "from configs.config import ConfigEnv, ConfigPath\n",
    "from knowledge_graph.connection import Neo4jConnection\n",
    "from llms.llm import ChatModel\n",
    "from data_collection.reader import BioASQDataReader\n",
    "from tqdm import tqdm\n",
    "from langchain_neo4j import Neo4jVector\n",
    "from retrieval_techniques.similarity_search import SimilaritySearchRetriever"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'question': 'What is the implication of histone lysine methylation in medulloblastoma?',\n",
       " 'answer': 'Aberrant patterns of H3K4, H3K9, and H3K27 histone lysine methylation were shown to result in histone code alterations, which induce changes in gene expression, and affect the proliferation rate of cells in medulloblastoma.',\n",
       " 'id': 1682,\n",
       " 'articles': [{'pmid': '23179372',\n",
       "   'title': 'OTX2 sustains a bivalent-like state of OTX2-bound promoters in medulloblastoma by maintaining their H3K27me3 levels.',\n",
       "   'abstract': 'Recent studies showed frequent mutations in histone H3 lysine 27 (H3K27) demethylases in medulloblastomas of Group 3 and Group 4, suggesting a role for H3K27 methylation in these tumors. Indeed, trimethylated H3K27 (H3K27me3) levels were shown to be higher in Group 3 and 4 tumors compared to WNT and SHH medulloblastomas, also in tumors without detectable mutations in demethylases. Here, we report that polycomb genes, required for H3K27 methylation, are consistently upregulated in Group 3 and 4 tumors. These tumors show high expression of the homeobox transcription factor OTX2. Silencing of OTX2 in D425 medulloblastoma cells resulted in downregulation of polycomb genes such as EZH2, EED, SUZ12 and RBBP4 and upregulation of H3K27 demethylases KDM6A, KDM6B, JARID2 and KDM7A. This was accompanied by decreased H3K27me3 and increased H3K27me1 levels in promoter regions. Strikingly, the decrease of H3K27me3 was most prominent in promoters that bind OTX2. OTX2-bound promoters showed high levels of the H3K4me3 and H3K9ac activation marks and intermediate levels of the H3K27me3 inactivation mark, reminiscent of a bivalent modification. After silencing of OTX2, H3K27me3 levels strongly dropped, but H3K4me3 and H3K9ac levels remained high. OTX2-bound bivalent genes showed high expression levels in D425, but the expression of most of these genes did not change after OTX2 silencing and loss of the H3K27me3 mark. Maintaining promoters in a bivalent state by sustaining H3K27 trimethylation therefore seems to be an important function of OTX2 in medulloblastoma, while other transcription factors might regulate the actual expression levels of these genes.',\n",
       "   'mesh_terms': ['Cell Line, Tumor',\n",
       "    'Chromatin Immunoprecipitation',\n",
       "    'Gene Expression Profiling',\n",
       "    'Gene Expression Regulation, Neoplastic',\n",
       "    'Histones',\n",
       "    'Humans',\n",
       "    'Medulloblastoma',\n",
       "    'Methylation',\n",
       "    'Oligonucleotide Array Sequence Analysis',\n",
       "    'Otx Transcription Factors',\n",
       "    'Polycomb-Group Proteins',\n",
       "    'Promoter Regions, Genetic',\n",
       "    'RNA, Messenger']},\n",
       "  {'pmid': '19270706',\n",
       "   'title': 'Multiple recurrent genetic events converge on control of histone lysine methylation in medulloblastoma.',\n",
       "   'abstract': 'We used high-resolution SNP genotyping to identify regions of genomic gain and loss in the genomes of 212 medulloblastomas, malignant pediatric brain tumors. We found focal amplifications of 15 known oncogenes and focal deletions of 20 known tumor suppressor genes (TSG), most not previously implicated in medulloblastoma. Notably, we identified previously unknown amplifications and homozygous deletions, including recurrent, mutually exclusive, highly focal genetic events in genes targeting histone lysine methylation, particularly that of histone 3, lysine 9 (H3K9). Post-translational modification of histone proteins is critical for regulation of gene expression, can participate in determination of stem cell fates and has been implicated in carcinogenesis. Consistent with our genetic data, restoration of expression of genes controlling H3K9 methylation greatly diminishes proliferation of medulloblastoma in vitro. Copy number aberrations of genes with critical roles in writing, reading, removing and blocking the state of histone lysine methylation, particularly at H3K9, suggest that defective control of the histone code contributes to the pathogenesis of medulloblastoma.',\n",
       "   'mesh_terms': ['Cerebellar Neoplasms',\n",
       "    'Gene Amplification',\n",
       "    'Gene Deletion',\n",
       "    'Genes, Tumor Suppressor',\n",
       "    'Genome, Human',\n",
       "    'Histone-Lysine N-Methyltransferase',\n",
       "    'Histones',\n",
       "    'Humans',\n",
       "    'Lysine',\n",
       "    'Medulloblastoma',\n",
       "    'Polymorphism, Single Nucleotide',\n",
       "    'Protein Processing, Post-Translational',\n",
       "    'Sequence Deletion']},\n",
       "  {'pmid': '23184418',\n",
       "   'title': 'Aberrant patterns of H3K4 and H3K27 histone lysine methylation occur across subgroups in medulloblastoma.',\n",
       "   'abstract': 'Recent sequencing efforts have described the mutational landscape of the pediatric brain tumor medulloblastoma. Although MLL2 is among the most frequent somatic single nucleotide variants (SNV), the clinical and biological significance of these mutations remains uncharacterized. Through targeted re-sequencing, we identified mutations of MLL2 in 8 % (14/175) of MBs, the majority of which were loss of function. Notably, we also report mutations affecting the MLL2-binding partner KDM6A, in 4 % (7/175) of tumors. While MLL2 mutations were independent of age, gender, histological subtype, M-stage or molecular subgroup, KDM6A mutations were most commonly identified in Group 4 MBs, and were mutually exclusive with MLL2 mutations. Immunohistochemical staining for H3K4me3 and H3K27me3, the chromatin effectors of MLL2 and KDM6A activity, respectively, demonstrated alterations of the histone code in 24 % (53/220) of MBs across all subgroups. Correlating these MLL2- and KDM6A-driven histone marks with prognosis, we identified populations of MB with improved (K4+/K27-) and dismal (K4-/K27-) outcomes, observed primarily within Group 3 and 4 MBs. Group 3 and 4 MBs demonstrate somatic copy number aberrations, and transcriptional profiles that converge on modifiers of H3K27-methylation (EZH2, KDM6A, KDM6B), leading to silencing of PRC2-target genes. As PRC2-mediated aberrant methylation of H3K27 has recently been targeted for therapy in other diseases, it represents an actionable target for a substantial percentage of medulloblastoma patients with aggressive forms of the disease.',\n",
       "   'mesh_terms': ['Base Sequence',\n",
       "    'Cerebellar Neoplasms',\n",
       "    'Cohort Studies',\n",
       "    'DNA-Binding Proteins',\n",
       "    'Female',\n",
       "    'Genetic Predisposition to Disease',\n",
       "    'Genome',\n",
       "    'Histone Demethylases',\n",
       "    'Histone Methyltransferases',\n",
       "    'Histone-Lysine N-Methyltransferase',\n",
       "    'Humans',\n",
       "    'Lysine',\n",
       "    'Male',\n",
       "    'Medulloblastoma',\n",
       "    'Methylation',\n",
       "    'Mutation',\n",
       "    'Neoplasm Proteins',\n",
       "    'Nuclear Proteins',\n",
       "    'Polymorphism, Single Nucleotide']}]}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "graph_data = read_json_file(file_path=os.path.join(path=os.path.join(ConfigPath.RAW_DATA_DIR, \"bioasq_graph_data.json\")))\n",
    "len(graph_data)\n",
    "graph_data[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'question': 'What is Trypan blue used for?',\n",
       "  'answer': 'Trypan blue is used in the \"trypan blue exclusion assay\" for assessing cell viability/cell death.',\n",
       "  'id': 1159,\n",
       "  'articles': [{'pmid': '24123008',\n",
       "    'title': 'Conjugated linoleate reduces prostate cancer viability whereas the effects of oleate and stearate are cell line-dependent.',\n",
       "    'abstract': 'BACKGROUND: In this study, responses to fatty acid treatments in commonly used prostate cancer cell culture models and variability of gene expression between them were determined. MATERIALS AND METHODS: PC3, DU145, LNCaP, VCaP and PNT2 cells were treated with 100 muM of either oleate, stearate or conjugated linoleate. Cell proliferation and viability were assessed using trypan blue and 3-(4, 5-dimethylthiazol-2-yl)-2, 5-diphenyltetrazolium bromide (MTT) assay respectively. Gene expression was measured using real-time polymerase chain reaction (PCR). RESULTS: Conjugated linoleic acid reduced cell proliferation and viability in all prostate cancer cell lines, whilst the effects of oleic and stearic acid on proliferation were found to be cell line-dependent. A reduction in gene expression of fatty acid desaturases was observed in prostate cancer cell lines compared to normal prostate cells. CONCLUSION: Differential responses of the cell lines investigated here to fatty acid treatment suggest that multiple prostate cancer cell line models should be used when designing experiments aimed at examining lipid metabolism in prostate cancer.',\n",
       "    'mesh_terms': ['Acetyl-CoA Carboxylase',\n",
       "     'Antineoplastic Agents',\n",
       "     'Cell Line, Tumor',\n",
       "     'Cell Proliferation',\n",
       "     'Cell Survival',\n",
       "     'Drug Screening Assays, Antitumor',\n",
       "     'Fatty Acid Synthase, Type I',\n",
       "     'Gene Expression',\n",
       "     'Glycerol-3-Phosphate O-Acyltransferase',\n",
       "     'Humans',\n",
       "     'Linoleic Acids, Conjugated',\n",
       "     'Lipid Metabolism',\n",
       "     'Male',\n",
       "     'Oleic Acid',\n",
       "     'PPAR gamma',\n",
       "     'Prostatic Neoplasms',\n",
       "     'Stearic Acids']},\n",
       "   {'pmid': '24228750',\n",
       "    'title': 'Coating-dependent induction of cytotoxicity and genotoxicity of iron oxide nanoparticles.',\n",
       "    'abstract': 'Surface coatings of nanoparticles (NPs) are known to influence advantageous features of NPs as well as potential toxicity. Iron oxide (Fe3O4) NPs are applied for both medical diagnostics and targeted drug delivery. We investigated the potential cytotoxicity and genotoxicity of uncoated iron oxide (U-Fe3O4) NPs in comparison with oleate-coated iron oxide (OC-Fe3O4) NPs. Testing was performed in vitro in human lymphoblastoid TK6 cells and in primary human blood cells. For cytotoxicity testing, relative growth activity, trypan blue exclusion, (3)H-thymidine incorporation and cytokinesis-block proliferation index were assessed. Genotoxicity was evaluated by the alkaline comet assay for detection of strand breaks and oxidized purines. Particle characterization was performed in the culture medium. Cellular uptake, morphology and pathology were evaluated by electron microscopy. U-Fe3O4 NPs were found not to be cytotoxic (considering interference of NPs with proliferation test) or genotoxic under our experimental conditions. In contrast, OC-Fe3O4 NPs were cytotoxic in a dose-dependent manner, and also induced DNA damage, indicating genotoxic potential. Intrinsic properties of sodium oleate were excluded as a cause of the toxic effect. Electron microscopy data were consistent with the cytotoxicity results. Coating clearly changed the behaviour and cellular uptake of the NPs, inducing pathological morphological changes in the cells.',\n",
       "    'mesh_terms': ['Cell Line',\n",
       "     'Cell Proliferation',\n",
       "     'Comet Assay',\n",
       "     'Cytotoxins',\n",
       "     'DNA Damage',\n",
       "     'Ferric Compounds',\n",
       "     'Humans',\n",
       "     'Magnetite Nanoparticles',\n",
       "     'Mutagens',\n",
       "     'Surface Properties']},\n",
       "   {'pmid': '24127887',\n",
       "    'title': 'Dental pulp stem cells immobilized in alginate microspheres for applications in bone tissue engineering.',\n",
       "    'abstract': 'AIM: To immobilize dental pulp stem cells (DPSC) in alginate microspheres and to determine cell viability, proliferation, stem cell characteristics and osteogenic potential of the immobilized DPSCs. METHODOLOGY: Human DPSCs isolated from the dental pulp were immobilized in 1% w/v alginate microspheres. Viability and proliferation of immobilized DPSCs were determined by trypan blue and MTT assay, respectively. Stem cell characteristics of DPSCs post immobilization were verified by labelling the cells with CD73 and CD90. Osteogenic potential of immobilized DPSCs was assessed by the presence of osteocalcin. Alizarin red staining and O-cresolphthalein complexone method confirmed and quantified calcium deposition. A final reverse transcriptase PCR evaluated the expression of osteogenic markers - ALP, Runx-2 and OCN. RESULTS: More than 80% of immobilized DPSCs were viable throughout the 3-week study. Proliferation appeared controlled and consistent unlike DPSCs in the control group. Presence of CD73 and CD90 markers confirmed the stem cell nature of immobilized DPSCs. The presence of osteocalcin, an osteoblastic marker, was confirmed in the microspheres on day 21. Mineralization assays showed high calcium deposition indicating elevated osteogenic potential of immobilized DPSCs. Osteogenic genes- ALP, Runx-2 and OCN were also upregulated in immobilized DPSCs. Surprisingly, immobilized DPSCs in the control group cultured in conventional stem cell media showed upregulation of osteogenic genes and expressed osteocalcin. CONCLUSION: Dental pulp stem cells immobilized in alginate hydrogels exhibit enhanced osteogenic potential while maintaining high cell viability both of which are fundamental for bone tissue regeneration.',\n",
       "    'mesh_terms': ['Alginates',\n",
       "     'Bone Development',\n",
       "     'Cell Proliferation',\n",
       "     'Cells, Cultured',\n",
       "     'Dental Pulp',\n",
       "     'Glucuronic Acid',\n",
       "     'Hexuronic Acids',\n",
       "     'Humans',\n",
       "     'Microspheres',\n",
       "     'Reverse Transcriptase Polymerase Chain Reaction',\n",
       "     'Stem Cells',\n",
       "     'Tissue Engineering']},\n",
       "   {'pmid': '24231574',\n",
       "    'title': 'Amniotic fluid for ex vivo skin preservation: a comparative study of tissue preservation solutions.',\n",
       "    'abstract': \"Ex vivo skin preservation is important for skin banks, burn centers, and in research; however, the optimal preservation solution is not known. Human amniotic fluid (HAF), in addition to its role in fetal wound healing, has promise as an effective and readily available preservation solution. The purpose of this study was to compare the efficacy of several solutions, including HAF, in full-thickness skin preservation. Human amniotic fluid was obtained from patients undergoing amniocentesis. Full-thickness skin obtained during abdominoplasty was divided into 1-cm(2) samples. These specimens were preserved in either saline, HAF from a single patient, pooled HAF, University of Wisconsin solution, or custodial histidine-tryptophan-ketoglutarate solution at 4 degrees C. There were 5 samples in each group. Specimens were examined for keratinocyte survival at 7, 14, 21, 28, and 35 days using the trypan blue assay. The first 200 cells identified were counted to calculate the degree of cell death. Comparisons were made between the groups, and a multivariable repeated-measures analysis was performed to determine statistical significance, which was defined as P < 0.05. The individual and pooled HAF showed greater keratinocyte survival than the other solutions, at each time point. The difference was statistically significant (P < 0.05) when compared to saline and custodial solution but not statistically significant when compared to University of Wisconsin solution. When used as a tissue preservation solution, HAF produces superior keratinocyte survival than several commercially available solutions. Amniotic fluid, nature's tissue preservation solution, may be an inexpensive, readily available alternative for use in skin banks, burn centers, and research.\",\n",
       "    'mesh_terms': ['Amniotic Fluid',\n",
       "     'Cell Survival',\n",
       "     'Glucose',\n",
       "     'Humans',\n",
       "     'In Vitro Techniques',\n",
       "     'Keratinocytes',\n",
       "     'Mannitol',\n",
       "     'Multivariate Analysis',\n",
       "     'Organ Preservation Solutions',\n",
       "     'Potassium Chloride',\n",
       "     'Procaine',\n",
       "     'Skin',\n",
       "     'Sodium Chloride']},\n",
       "   {'pmid': '24228508',\n",
       "    'title': '[Connective tissue growth factor induced differentiation of placenta mesenchymal stem cell into dermal fibroblast].',\n",
       "    'abstract': 'OBJECTIVE: To investigate the possibility of placenta mesenchymal stem cells (PMSCs) differentiation into dermal fibroblast, and the potency of PMSCs used in cutaneous wound healing and stored as seed cells. METHODS: Enzyme digestion method was used to obtain PMSCs, and PMSCs were amplified after culture in vitro. Flow cytometry assay, osteogenic and adipogenic differentiation were done for MSCs identification. The induction medium composed of DMEM/F12 + 50 microg/ml VC + 100 ng/ml connective tissue growth factor (CTGF) was added into the 24-well plate for 16 days induction period. Pictures were taken to record morphologic change. Immunofluorescence tests were performed to detect Vimentin, FSP-1, collagen I , collagen III, desmin and laminin expression before and after induction. At the same time osteogenic and adipogenic differentiation were used to assay the differentiation ability change after induction. The induced dermal fibroblasts were frozen in liquid nitrogen and recovery and trypan blue was used to detect cell viability. RESULTS: After CTGF induction, PMSCs got obvious fibroblasts morphology, the protein level of Vimentin, FSP-1, collagen I, collagen III and Laminin increased, PMSCs started to express Desmin, the dermal fibroblasts specific proteins, and osteogenic and adipogenic differentiation ability was diminished. PMSCs were successfully induced into dermal fibroblasts, and these induced cells could get a high cell viability ( more than 90% ) after recovery. CONCLUSIONS: PMSCs could be induced into dermal fibroblasts by CTGF in vitro. PMSCs have the potential application in skin wound healing, and can be used as seed cells of dermal fibroblasts.',\n",
       "    'mesh_terms': ['Cell Differentiation',\n",
       "     'Cells, Cultured',\n",
       "     'Connective Tissue Growth Factor',\n",
       "     'Female',\n",
       "     'Fibroblasts',\n",
       "     'Humans',\n",
       "     'Mesenchymal Stem Cells',\n",
       "     'Placenta',\n",
       "     'Pregnancy']},\n",
       "   {'pmid': '24289578',\n",
       "    'title': 'Fenugreek induced apoptosis in breast cancer MCF-7 cells mediated independently by fas receptor change.',\n",
       "    'abstract': 'Trigonella foenum in graecum (Fenugreek) is a traditional herbal plant used to treat disorders like diabetes, high cholesterol, wounds, inflammation, gastrointestinal ailments, and it is believed to have anti-tumor properties, although the mechanisms for the activity remain to be elucidated. In this study, we prepared a methanol extract from Fenugreek whole plants and investigated the mechanism involved in its growth-inhibitory effect on MCF- 7 human breast cancer cells. Apoptosis of MCF-7 cells was evidenced by investigating trypan blue exclusion, TUNEL and Caspase 3, 8, 9, p53, FADD, Bax and Bak by real-time PCR assays inducing activities, in the presence of FME at 65 mug/mL for 24 and 48 hours. FME induced apoptosis was mediated by the death receptor pathway as demonstrated by the increased level of Fas receptor expression after FME treatment. However, such change was found to be absent in Caspase 3, 8, 9, p53, FADD, Bax and Bak, which was confirmed by a time-dependent and dose-dependent manner. In summary, these data demonstrate that at least 90% of FME induced apoptosis in breast cell is mediated by Fas receptor-independently of either FADD, Caspase 8 or 3, as well as p53 interdependently.',\n",
       "    'mesh_terms': ['Antineoplastic Agents',\n",
       "     'Apoptosis',\n",
       "     'Breast Neoplasms',\n",
       "     'Caspases',\n",
       "     'Cell Line, Tumor',\n",
       "     'Fas-Associated Death Domain Protein',\n",
       "     'Humans',\n",
       "     'MCF-7 Cells',\n",
       "     'Plant Extracts',\n",
       "     'Plants, Medicinal',\n",
       "     'Trigonella',\n",
       "     'Tumor Suppressor Protein p53',\n",
       "     'bcl-2 Homologous Antagonist-Killer Protein',\n",
       "     'bcl-2-Associated X Protein',\n",
       "     'fas Receptor']},\n",
       "   {'pmid': '24244400',\n",
       "    'title': 'Pretreatment with apoaequorin protects hippocampal CA1 neurons from oxygen-glucose deprivation.',\n",
       "    'abstract': 'Ischemic stroke affects  approximately 795,000 people each year in the U.S., which results in an estimated annual cost of $73.7 billion. Calcium is pivotal in a variety of neuronal signaling cascades, however, during ischemia, excess calcium influx can trigger excitotoxic cell death. Calcium binding proteins help neurons regulate/buffer intracellular calcium levels during ischemia. Aequorin is a calcium binding protein isolated from the jellyfish Aequorea victoria, and has been used for years as a calcium indicator, but little is known about its neuroprotective properties. The present study used an in vitro rat brain slice preparation to test the hypothesis that an intra-hippocampal infusion of apoaequorin (the calcium binding component of aequorin) protects neurons from ischemic cell death. Bilaterally cannulated rats received an apoaequorin infusion in one hemisphere and vehicle control in the other. Hippocampal slices were then prepared and subjected to 5 minutes of oxygen-glucose deprivation (OGD), and cell death was assayed by trypan blue exclusion. Apoaequorin dose-dependently protected neurons from OGD--doses of 1% and 4% (but not 0.4%) significantly decreased the number of trypan blue-labeled neurons. This effect was also time dependent, lasting up to 48 hours. This time dependent effect was paralleled by changes in cytokine and chemokine expression, indicating that apoaequorin may protect neurons via a neuroimmunomodulatory mechanism. These data support the hypothesis that pretreatment with apoaequorin protects neurons against ischemic cell death, and may be an effective neurotherapeutic.',\n",
       "    'mesh_terms': ['Aequorin',\n",
       "     'Animals',\n",
       "     'Apoproteins',\n",
       "     'Brain Ischemia',\n",
       "     'CA1 Region, Hippocampal',\n",
       "     'Cell Death',\n",
       "     'Glucose',\n",
       "     'Humans',\n",
       "     'Neurons',\n",
       "     'Oxygen',\n",
       "     'Oxygen Consumption',\n",
       "     'Rats',\n",
       "     'Recombinant Proteins']},\n",
       "   {'pmid': '24160177',\n",
       "    'title': 'Knockdown of autophagy-related protein 5, ATG5, decreases oxidative stress and has an opposing effect on camptothecin-induced cytotoxicity in osteosarcoma cells.',\n",
       "    'abstract': 'BACKGROUND: Autophagy induction can increase or decrease anticancer drug efficacy. Anticancer drug-induced autophagy induction is poorly characterized in osteosarcoma (OS). In this study, we investigated the impact of autophagy inhibition on camptothecin (CPT)-induced cytotoxicity in OS. METHODS: Autophagy-inhibited DLM8 and K7M3 metastatic murine OS cell lines were generated by infection with lentiviral shRNA directed against the essential autophagy protein ATG5. Knockdown of ATG5 protein expression and inhibition of autophagy was confirmed by immunoblot of ATG5 and LC3II proteins, respectively. Metabolic activity was determined by MTT assay and cell viability was determined by trypan blue exclusion. Acridine orange staining and immunoblotting for LC3II protein expression were used to determine autophagy induction. Oxidative stress was assessed by staining cells with HE and DCFH-DA followed by flow cytometry analysis. Mitochondrial membrane potential was determined by staining cells with TMRE followed by flow cytometry analysis. Immunoblotting was used to detect caspase activation, Parp cleavage and p53 phosphorylation. RESULTS: Autophagy inhibition caused a greater deficit in metabolic activity and cell growth in K7M3 cells compared to DLM8 cells. K7M3 cells exhibited higher basal autophagy levels than DLM8 cells and non-transformed murine MCT3 osteoblasts. Autophagy inhibition did not affect CPT-induced DNA damage. Autophagy inhibition decreased CPT-induced cell death in DLM8 cells while increasing CPT-induced cell death in K7M3 cells. Autophagy inhibition reduced CPT-induced mitochondrial damage and CPT-induced caspase activation in DLM8 cells. Buthionine sulfoximine (BSO)-induced cell death was greater in autophagy-competent DLM8 cells and was reversed by antioxidant pretreatment. Camptothecin-induced and BSO-induced autophagy induction was also reversed by antioxidant pretreatment. Significantly, autophagy inhibition not only reduced CPT-induced oxidative stress but also reduced basal oxidative stress. CONCLUSIONS: The results of this study indicate that autophagy inhibition can have an opposing effect on CPT-induced cytotoxicity within OS. The cytoprotective mechanism of autophagy inhibition observed in DLM8 cells involves reduced CPT-induced oxidative stress and not reduced DNA damage. Our results also reveal the novel finding that knockdown of ATG5 protein reduces both basal oxidative stress and drug-induced oxidative stress.',\n",
       "    'mesh_terms': ['Animals',\n",
       "     'Antineoplastic Agents, Phytogenic',\n",
       "     'Autophagy',\n",
       "     'Autophagy-Related Protein 5',\n",
       "     'Bone Neoplasms',\n",
       "     'Buthionine Sulfoximine',\n",
       "     'Camptothecin',\n",
       "     'Caspases',\n",
       "     'Cell Line, Tumor',\n",
       "     'Cell Survival',\n",
       "     'Enzyme Activation',\n",
       "     'Gene Knockdown Techniques',\n",
       "     'Membrane Potential, Mitochondrial',\n",
       "     'Mice',\n",
       "     'Microtubule-Associated Proteins',\n",
       "     'Osteosarcoma',\n",
       "     'Oxidative Stress']},\n",
       "   None,\n",
       "   {'pmid': '24300339',\n",
       "    'title': 'Exposure to low- vs iso-osmolar contrast agents reduces NADPH-dependent reactive oxygen species generation in a cellular model of renal injury.',\n",
       "    'abstract': 'Contrast-induced nephropathy represents the third cause of hospital-acquired acute renal failure. This study investigated the effects of low- vs iso-osmolar contrast medium (CM) exposure on NADPH-dependent reactive oxygen species (ROS) generation by tubular cells. X-ray attenuation of iohexol, iopamidol, and iodixanol was assessed at equimolar iodine concentrations and their effects on human renal proximal tubular cells (PTCs) were evaluated with equally attenuating solutions of each CM. Cytotoxicity, apoptosis, and necrosis were investigated by trypan blue exclusion, MTT assay, and annexin V/propidium iodide assay, respectively. ROS production was assessed by DCF assay, NADPH oxidase activity by the lucigenin-enhanced chemiluminescence method, and Nox4 expression by immunoblot. Yielding the same X-ray attenuation, CM cytotoxicity was assessed in PTCs at equimolar iodine concentrations. More necrosis was present after incubation with iohexol and iopamidol than after incubation with equal concentrations of iodixanol. Iohexol and iodixanol at low iodine concentrations induced less cytotoxicity than iopamidol. Moreover, both iohexol and iopamidol induced more apoptosis than iodixanol, with a dose-dependent effect. ROS generation was significantly higher with iopamidol and iohexol compared to iodixanol. NADPH oxidase activity and Nox4 protein expression significantly increased after exposure to iopamidol and iohexol, with a dose-dependent effect, compared with iodixanol. CM-induced Nox4 expression and activity depended upon Src activation. In conclusion, at angiographic concentrations, iodixanol induces fewer cytotoxic effects on cultured tubular cells than iohexol and iopamidol along with a lower induction of Nox4-dependent ROS generation. This enzyme may, thus, represent a potential therapeutic target to prevent iodinated CM-related oxidative stress.',\n",
       "    'mesh_terms': ['Acute Kidney Injury',\n",
       "     'Apoptosis',\n",
       "     'Contrast Media',\n",
       "     'Humans',\n",
       "     'Iohexol',\n",
       "     'Iopamidol',\n",
       "     'Kidney',\n",
       "     'NADP',\n",
       "     'NADPH Oxidase 4',\n",
       "     'NADPH Oxidases',\n",
       "     'Necrosis',\n",
       "     'Radiography',\n",
       "     'Reactive Oxygen Species',\n",
       "     'Triiodobenzoic Acids']},\n",
       "   {'pmid': '24195509',\n",
       "    'title': 'INCB018424 induces apoptotic cell death through the suppression of pJAK1 in human colon cancer cells.',\n",
       "    'abstract': 'Janus kinase (JAK) is one of the main upstream activators of signal transducers and activators of transcription (STAT) that are constitutively activated in various malignancies and are associated with cell growth, survival, and carcinogenesis. Here, we investigated the role of JAKs in colorectal cancer in order to develop effective therapeutic targets for INCB018424, which is the first JAK1/2 inhibitor to be approved by FDA. After examining the basal expression levels of phospho-JAK1 and phospho-JAK2, we measured the effects of INCB018424 on the phosphorylation of JAK1/2 using western blot analysis. Cell viability was determined using the trypan blue exclusion assay. The cell death mechanism was identified by the activation of caspase 3 using western blot and annexin V staining. The basal levels of phospho-JAK1 and phospho-JAK2 were cancer cell type dependent. Colorectal cancer cell lines that phosphorylate both JAK1 and JAK2 include DLD-1 and RKO. INCB018424 inactivates both JAK1 and JAK2 in DLD-1 cells but inactivates only JAK1 in RKO cells. Cell death was proportional to the inactivation of JAK1 but not JAK2. INCB018424 causes caspase-dependent cell death, which is prevented by treatment with z-VAD. The inhibition of JAK1 phosphorylation seemed sufficient to allow INCB018424-mediated apoptosis. JAK1 is a key molecule that is involved in colon cancer cell survival and the inhibition of JAK1 by INCB01424 results in caspase-dependent apoptosis in colorectal cancer cells. The use of selective JAK1 inhibitors could be an attractive therapy against colorectal cancer, but further clinical investigations are needed to test this possibility.',\n",
       "    'mesh_terms': ['Apoptosis',\n",
       "     'Cell Line, Tumor',\n",
       "     'Colonic Neoplasms',\n",
       "     'Humans',\n",
       "     'Janus Kinase 1',\n",
       "     'Nitriles',\n",
       "     'Phosphorylation',\n",
       "     'Pyrazoles',\n",
       "     'Pyrimidines',\n",
       "     'STAT Transcription Factors',\n",
       "     'Signal Transduction']},\n",
       "   {'pmid': '24316855',\n",
       "    'title': 'An outcome analysis and long-term viability of cryopreserved cultured epidermal allografts: assessment of the conservation of transplantable human skin allografts.',\n",
       "    'abstract': 'PURPOSE: To assess the viability of cultured epithelium and preserved by freezing for periods varying from one month to one year. METHODS: Samples of cultured epithelium were incubated in cryoprotectant medium (Group A), packed in aluminum envelopes and packed in polystyrene boxes. The boxes were subjected to a temperature of-70  masculineC. After freezing for a period of time ranging from one to 12 months, cultured epithelial samples were assessed for their viability by vital staining (Trypan blue) and metabolic analysis based on glucose consumption and lactate production. Samples of not frozen cultured epithelium (Group B) were also tested for viability and the results obtained were used as comparison parameter for the variation of viability. RESULTS: Statistical analysis between the group A and B indicate that the mean age of the donors (p=0.51) and the culture time (p=1.18) showed no statistical difference. In 30 days we obtained 37% of the original viability of cultured epithelium, 25% at six months and one year, less than 15%. This trend was confirmed statistically with a reduction of approximately 1.8% of the original viability epithelium cultured every 30 days of storage. In the analysis by lactate production, similar results were observed. In the analysis by the glucose consumption results were not significant. The viability indices show statistically significant difference between the group A and B (p<0.0001). CONCLUSIONS: Although cryopreserved cultured epithelium showed significant reduction of viability, all samples remained viable. It was also found that the viability of cryopreserved cultured epithelial decreased as a function of storage time.',\n",
       "    'mesh_terms': ['Adult',\n",
       "     'Allografts',\n",
       "     'Cell Survival',\n",
       "     'Cells, Cultured',\n",
       "     'Cryopreservation',\n",
       "     'Cryoprotective Agents',\n",
       "     'Culture Media',\n",
       "     'Epithelium',\n",
       "     'Female',\n",
       "     'Humans',\n",
       "     'Male',\n",
       "     'Reference Values',\n",
       "     'Reproducibility of Results',\n",
       "     'Skin',\n",
       "     'Statistics, Nonparametric',\n",
       "     'Time Factors',\n",
       "     'Tissue Survival',\n",
       "     'Young Adult']},\n",
       "   {'pmid': '24304568',\n",
       "    'title': 'Pterostilbene induces accumulation of autophagic vacuoles followed by cell death in HL60 human leukemia cells.',\n",
       "    'abstract': 'Pterostilbene, a naturally occurring structural analog of resveratrol, has been reported to exert antiproliferative and proapoptotic effects in various cancer types. Recently, it has been demonstrated to induce both autophagy and apoptosis in human bladder and breast cancer cell lines. The aim of this study was to evaluate the effects of pterostilbene on HL60 human leukemia cells. Cell morphology was examined using confocal and electron microscopy. Cell viability was determined by MTT, neutral red uptake and trypan blue exclusion assays. LC3 processing was studied based on Western blotting and immunofluorescence analyses. Flow cytometry was used to study cell cycle distribution, phosphatidylserine externalization, caspase activation, disruption of mitochondrial membrane potential and intracellular production of reactive oxygen species. DNA degradation was examined by gel electrophoresis. We found that treatment of HL60 cells with pterostilbene at the IC90 concentration resulted in the G0/G1 cell cycle arrest. Pterostilbene induced conversion of cytosolic LC3-I to membrane-bound LC3-II and accumulation of large LC3-positive vacuolar structures. Pterostilbene also led to phosphatidylserine externalization, internucleosomal DNA fragmentation, caspase activation and disruption of mitochondrial membrane potential. Moreover, it did not induce oxidative stress. Our results suggest that pterostilbene induces accumulation of autophagic vacuoles followed by cell death in HL60 cells.',\n",
       "    'mesh_terms': ['Antineoplastic Agents',\n",
       "     'Autophagy',\n",
       "     'Caspases',\n",
       "     'Cell Cycle',\n",
       "     'Cell Death',\n",
       "     'Cell Survival',\n",
       "     'DNA Fragmentation',\n",
       "     'HL-60 Cells',\n",
       "     'Humans',\n",
       "     'Leukemia',\n",
       "     'Membrane Potential, Mitochondrial',\n",
       "     'Microtubule-Associated Proteins',\n",
       "     'Reactive Oxygen Species',\n",
       "     'Stilbenes',\n",
       "     'Vacuoles']},\n",
       "   {'pmid': '24236478',\n",
       "    'title': 'Assessment of genomic damage and repair on human lymphocytes by paint thinner in vitro.',\n",
       "    'abstract': 'Paint thinners are organic-solvent complex mixtures frequently used by car painters around the world in industries and shops. Some studies have revealed the oxidative effect induced by thinner inhalation; however, its genotoxic effect is poorly studied. The aim of this study was to assess the cytotoxicity, genomic damage and DNA repair in vitro induced by commercial paint thinner 0.14 in human lymphocytes. Cytotoxicity was determined by cell-viability analysis with trypan blue after 4 h treatment with different thinner concentrations (0.025 to 1.2 microL/mL). Genomic damage was evaluated by means of the alkaline single-cell gel electrophoresis (SCGE; pH > 13) in treated cultures after 1 h with three low-cytotoxic thinner concentrations (0.05, 0.1 and 0.2 microL/mL). In order to evaluate the genomic DNA repair, one set of SCGE slides was prepared immediately after treatment, and another one was prepared after 4 h of liquid-holding recovery. A significant level of cytotoxicity was observed over the entire concentration range of paint thinner in lymphocytes (F = 175.98; p </= 0.001). In the SCGE % tail DNA assessment, a significant increase of lymphocyte genomic damage was evidenced (F = 72.32; p < 0.001). In addition, we found a significant decrease in the % tail DNA in thinner-treated cells after liquid-holding recovery period (all p < 0.05), demonstrating that DNA primary lesions induced by low-cytotoxic thinner concentrations are efficiently repaired. In conclusion, thinner components induce cytotoxicity and genomic damage in human lymphocytes under the study conditions, possibly by oxidative and alkylative DNA damage.',\n",
       "    'mesh_terms': ['Comet Assay',\n",
       "     'DNA Repair',\n",
       "     'Humans',\n",
       "     'In Vitro Techniques',\n",
       "     'Lymphocytes',\n",
       "     'Solvents']},\n",
       "   {'pmid': '24129092',\n",
       "    'title': 'VEGF depletion enhances bcr-abl-specific sensitivity of arsenic trioxide in chronic myelogenous leukemia.',\n",
       "    'abstract': 'The development of resistance to imatinib mesylate may partly depend on high bcr-abl expression levels or point mutation(s). Arsenic trioxide (ATO) has bcr-abl suppressing activity in vitro, without cross-resistance to imatinib. Meanwhile, bcr-abl also induces expression of vascular endothelial growth factor (VEGF), which is associated with tumor-related angiogenesis and is involved in chronic myelogenous leukemia (CML) pathogenesis. Here, we investigated ways to improve ATO activity in CML by modulating cellular VEGF levels. K562 and primary CML cells were transfected with a VEGF antisense sequence. Cell viability and survival were assessed using 3-(4,5-dimethylthiazol-2-yl-2,5-diphenyltetrazolium bromide and trypan blue exclusion assays. Apoptotic cells were detected by flow cytometry following annexin V and propidium iodide staining. The results showed that VEGF depletion effectively promotes enhanced ATO antileukemic activity by repressing bcr-abl protein levels. These data provide a rationale for the clinical development of optimized ATO-based regimens that incorporate VEGF modulator for CML treatment.',\n",
       "    'mesh_terms': ['Apoptosis',\n",
       "     'Arsenic Trioxide',\n",
       "     'Arsenicals',\n",
       "     'Benzamides',\n",
       "     'Cell Line, Tumor',\n",
       "     'Down-Regulation',\n",
       "     'Drug Resistance, Neoplasm',\n",
       "     'Fusion Proteins, bcr-abl',\n",
       "     'Genes, abl',\n",
       "     'Humans',\n",
       "     'Imatinib Mesylate',\n",
       "     'K562 Cells',\n",
       "     'Leukemia, Myelogenous, Chronic, BCR-ABL Positive',\n",
       "     'Oligodeoxyribonucleotides, Antisense',\n",
       "     'Oxides',\n",
       "     'Piperazines',\n",
       "     'Pyrimidines',\n",
       "     'Transfection',\n",
       "     'Tumor Cells, Cultured',\n",
       "     'Vascular Endothelial Growth Factor A']},\n",
       "   {'pmid': '24296136',\n",
       "    'title': 'Fecal water genotoxicity in healthy free-living young Italian people.',\n",
       "    'abstract': 'Dietary habit affects the composition of human feces thus determining intestinal environment and exposure of colon mucosa to risk factors. Fecal water (FW) citotoxicity and genotoxicity were investigated in 33 healthy young Italian people, as well as the relationship between genotoxicity and nutrient intake or microflora composition. Two fecal samples were collected at 2 weeks apart and 3-d dietary diary was recorded for each volunteer. Cytotoxicity was measured using the Trypan Blue Dye Exclusion assay and genotoxicity using the Comet Assay (alkaline single-cell electrophoresis). Fecal bifidobacteria, total microbial count and nutrient intakes were also assessed. High intra- and inter-variability in genotoxicity data and in bacteria counts were found. None of the FW samples were citotoxic, but 90% of FW samples were genotoxic. Seventy five percent indicated intermediate and 15% were highly genotoxic. There was a different sex-related distribution. Genotoxicity was positively correlated to the total lipid intake in females and to the bifidobacteria/total bacteria count ratio in male volunteers. These results demonstrate that the majority of FW samples isolated from free-living Italian people show intermediate level of genotoxicity and sustain a relation between this possible non-invasive marker of colorectal cancer risk with both dietary habits and colonic ecosystem.',\n",
       "    'mesh_terms': ['Caco-2 Cells',\n",
       "     'Diet',\n",
       "     'Feces',\n",
       "     'Female',\n",
       "     'Humans',\n",
       "     'Italy',\n",
       "     'Male',\n",
       "     'Mutagenicity Tests',\n",
       "     'Water']},\n",
       "   {'pmid': '24256980',\n",
       "    'title': 'Potential therapeutic role of Tridham in human hepatocellular carcinoma cell line through induction of p53 independent apoptosis.',\n",
       "    'abstract': 'BACKGROUND: Hepatocellular carcinoma (HCC) is the third leading cause of cancer deaths reported worldwide. The incidence is higher in Asia and Africa, where there is greater endemic prevalence of hepatitis B and C. The devastating outcome of cancer can be minimized only by the use of potent therapeutic agents. Tridham (TD) has been acknowledged since olden days for its wide spectrum of biological properties and was used by traditional practitioners of Siddha and other indigenous systems of medicine. The present study aims at investigating the mechanistic action of TD by assessing the antiproliferative and pro-apoptotic effects on human hepatocellular carcinoma cell line (Huh7). METHODS: Cell viability and apoptosis assay using MTT analysis and trypan blue staining, DAPI staining, DNA fragmentation, cell cycle analysis, mitochondrial membrane potential, real-time reverse transcription-polymerase chain reaction, western blotting and immunofluorescence staining were determined in Huh7 cells. RESULTS: Viability studies of TD treated Huh7 cells showed an inhibition in cell growth in time and dose dependent manner. Chromatin condensation, DNA fragmentation and apoptotic bodies, which are structural changes characteristic of apoptosis, were found following TD treatment of Huh7 cells. DAPI staining and agarose gel electrophoresis confirmed the induction of apoptosis by TD. Cell cycle analysis of Huh7 cells treated with TD exhibited a marked accumulation of cells in the sub-G1 phase of the cell cycle in a dose dependent manner. Immunofluorescent staining for Ki-67 showed a higher level of expression in untreated cells as compared to TD treated cells. We observed a significant loss in the mitochondrial membrane potential and the release of cytochrome c into the cytosol in TD treated cells. Down regulation of Bcl-2, up regulation of Bax and Bad as well as activation of caspases-3 and 9 were also observed. The p53 gene expression was found to be unaltered in TD treated cells. CONCLUSION: These results suggest that TD induces apoptosis of Huh7 cells through activation of Bax and triggered caspase cascade, independent of p53 function. This study throws light on the mechanistic action of TD in triggering apoptosis in Huh 7 cells.',\n",
       "    'mesh_terms': ['Apoptosis',\n",
       "     'Carcinoma, Hepatocellular',\n",
       "     'Cell Cycle',\n",
       "     'Cell Line, Tumor',\n",
       "     'Cell Nucleus',\n",
       "     'Cell Proliferation',\n",
       "     'Cell Survival',\n",
       "     'Cytochromes c',\n",
       "     'DNA Fragmentation',\n",
       "     'Dose-Response Relationship, Drug',\n",
       "     'Humans',\n",
       "     'Liver Neoplasms',\n",
       "     'Medicine, Ayurvedic',\n",
       "     'Membrane Potential, Mitochondrial',\n",
       "     'Plant Extracts',\n",
       "     'Tumor Suppressor Protein p53']},\n",
       "   {'pmid': '24279639',\n",
       "    'title': 'Activation of muscarinic receptors protects against retinal neurons damage and optic nerve degeneration in vitro and in vivo models.',\n",
       "    'abstract': 'AIMS: Muscarinic acetylcholine receptor agonist pilocarpine reduces intraocular pressure (IOP) of glaucoma mainly by stimulating ciliary muscle contraction and then increasing aqueous outflow. It is of our great interest to know whether pilocarpine has the additional properties of retinal neuroprotection independent of IOP lowering in vitro and in vivo models. METHODS: In rat primary retinal cultures, cell viability was measured using an MTT assay and the trypan blue exclusion method, respectively. Retinal ganglion cells (RGCs) were identified by immunofluorescence and quantified by flow cytometry. For the in vivo study, the retinal damage after retinal ischemia/reperfusion injury in rats was evaluated by histopathological study using hematoxylin and eosin staining, transmission electron microscopy, and immunohistochemical study on cleaved caspase-3, caspase-3, and ChAT. RESULTS: Pretreatment of pilocarpine attenuated glutamate-induced neurotoxicity of primary retinal neurons in a dose-dependent manner. Protection of pilocarpine in both retinal neurons and RGCs was largely abolished by the nonselective muscarinic receptor antagonist atropine and the M1-selective muscarinic receptor antagonist pirenzepine. After ischemia/reperfusion injury in retina, the inner retinal degeneration occurred including ganglion cell layer thinning and neuron lost, and the optic nerve underwent vacuolar changes. These degenerative changes were significantly lessened by topical application of 2% pilocarpine. In addition, the protective effect of pilocarpine on the ischemic rat retina was favorably reflected by downregulating the expression of activated apoptosis marker cleaved caspase-3 and caspase-3 and upregulating the expression of cholinergic cell marker ChAT. CONCLUSIONS: Taken together, this highlights pilocarpine through the activation of muscarinic receptors appear to afford significant protection against retinal neurons damage and optic nerve degeneration at clinically relevant concentrations. These data also further support muscarinic receptors as potential therapeutic neuroprotective targets in glaucoma.',\n",
       "    'mesh_terms': ['Animals',\n",
       "     'Animals, Newborn',\n",
       "     'Caspase 3',\n",
       "     'Cells, Cultured',\n",
       "     'Choline O-Acetyltransferase',\n",
       "     'Disease Models, Animal',\n",
       "     'Dose-Response Relationship, Drug',\n",
       "     'Excitatory Amino Acid Agonists',\n",
       "     'Glutamic Acid',\n",
       "     'Male',\n",
       "     'Muscarinic Agonists',\n",
       "     'Nerve Degeneration',\n",
       "     'Neurons',\n",
       "     'Optic Nerve',\n",
       "     'Pilocarpine',\n",
       "     'Rats',\n",
       "     'Rats, Sprague-Dawley',\n",
       "     'Receptors, Muscarinic',\n",
       "     'Reperfusion Injury',\n",
       "     'Retina']},\n",
       "   {'pmid': '24320727',\n",
       "    'title': 'Mitochondrial uncoupling protein 2 induces cell cycle arrest and necrotic cell death.',\n",
       "    'abstract': 'Uncoupling protein 2 (UCP2) is a mitochondrial membrane protein that regulates energy metabolism and reactive oxygen species (ROS) production. We generated mouse carboxy- and amino-terminal green fluorescent protein (GFP)-tagged UCP2 constructs to investigate the effect of UCP2 expression on cell proliferation and viability. UCP2-transfected Hepa 1-6 cells did not show reduced cellular adenosine triphosphate (ATP) but showed increased levels of glutathione. Flow cytometry analysis indicated that transfected cells were less proliferative than nontransfected controls, with most cells blocked at the G1 phase. The effect of UCP2 on cell cycle arrest could not be reversed by providing exogenous ATP or oxidant supply, and was not affected by the chemical uncoupler carbonyl cyanide-p-trifluoromethoxyphenylhydrazone (FCCP). However, this effect of UCP2 was augmented by treatment with genistein, a tyrosine kinase inhibitor, which by itself did not affect cell proliferation on control hepatocytes. Western blotting analysis revealed decreased expression levels of CDK6 but not CDK2 and D-type cyclins. Examination of cell viability in UCP2-transfected cells with Trypan Blue and Annexin-V staining revealed that UCP2 transfection led to significantly increased cell death. However, characteristics of apoptosis were absent in UCP2-transfected Hepa 1-6 cells, including lack of oligonucleosomal fragmentation (laddering) of chromosomal DNA, release of cytochrome c from mitochondria, and cleavage of caspase-3. In conclusion, our results indicate that UCP2 induces cell cycle arrest at G1 phase and causes nonapoptotic cell death, suggesting that UCP2 may act as a powerful influence on hepatic regeneration and cell death in the steatotic liver.',\n",
       "    'mesh_terms': ['Animals',\n",
       "     'Cell Cycle',\n",
       "     'Cell Death',\n",
       "     'Cell Proliferation',\n",
       "     'Cells, Cultured',\n",
       "     'Fatty Liver',\n",
       "     'Gene Expression Profiling',\n",
       "     'HEK293 Cells',\n",
       "     'HeLa Cells',\n",
       "     'Humans',\n",
       "     'Ion Channels',\n",
       "     'Liver',\n",
       "     'Liver Regeneration',\n",
       "     'Mice',\n",
       "     'Mitochondrial Proteins',\n",
       "     'Necrosis',\n",
       "     'Transfection',\n",
       "     'Uncoupling Protein 2']},\n",
       "   {'pmid': '24238300',\n",
       "    'title': 'The effect of temperature on the viability of human mesenchymal stem cells.',\n",
       "    'abstract': \"INTRODUCTION: Impaction allograft with cement is a common technique used in revision hip surgeries for the last 20 years. However, its clinical results are inconsistent. Recent studies have shown that mesenchymal stem cells (MSCs) seeded onto allograft can enhance bone formation. This in vitro study investigates whether the increase in temperature related to the polymerisation of bone cement will affect the viability of human MSCs. METHODS: The viability of human MSCs was measured after incubating them at temperatures of 38 degrees C, 48 degrees C and 58 degrees C; durations 45 seconds, 80 seconds and 150 seconds. A control group was kept at 37 degrees C and 5% carbon dioxide for the duration of the investigation (7 days). During the course of the study the human MSCs were analysed for cell metabolic activity using the alamarBlue assay, cell viability using both Trypan Blue dye exclusion and calcein staining under fluorescent microscopy, and necrosis and apoptosis using Annexin V and propidium iodide for flow cytometric analysis. A one-way analysis of variance with a priori Dunnett's test was used to indicate the differences between the treatment groups, when analysed against the control. This identified conditions with a significant difference in cell metabolic activity (alamarBlue) and cell viability (Trypan Blue). RESULTS: Results showed that cell metabolism was not severely affected up to 48 degrees C/150 seconds, while cells in the 58 degrees C group died. Similar results were shown using Trypan Blue and calcein analysis for cell viability. No significant difference in apoptosis and necrosis of the cells was observed when human MSCs treated at 48 degrees C/150 seconds were compared with the control group. CONCLUSIONS: The study suggests that human MSCs seeded onto allograft can be exposed to temperatures up to 48 degrees C for 150 seconds. Exposure to this temperature for this time period is unlikely to occur during impaction allograft surgery when cement is used. Therefore, in many situations, the addition of human MSCs to cemented impaction grafting may be carried out without detrimental effects to the cells. Furthermore, previous studies have shown that this can enhance new bone formation and repair the defects in revision situations.\",\n",
       "    'mesh_terms': ['Apoptosis',\n",
       "     'Bone Marrow Cells',\n",
       "     'Cell Survival',\n",
       "     'Cells, Cultured',\n",
       "     'Humans',\n",
       "     'Mesenchymal Stem Cells',\n",
       "     'Microscopy, Electron, Scanning',\n",
       "     'Necrosis',\n",
       "     'Temperature',\n",
       "     'Time Factors']},\n",
       "   {'pmid': '23354080',\n",
       "    'title': 'AKT serine/threonine protein kinase modulates baicalin-triggered autophagy in human bladder cancer T24 cells.',\n",
       "    'abstract': 'Baicalin is one of the major compounds in the traditional Chinese medicinal herb from Scutellaria baicalensis Georgi. We investigated the molecular mechanisms of cell autophagy induced by baicalin in human bladder cancer T24 cells. Baicalin inhibited cell survival as shown by MTT assay and increased cell death by trypan blue exclusion assay in a concentration-dependent manner. Baicalin did not induce apoptotic cell death in T24 cells by TUNEL and caspase-3 activity assay. Baicalin induced the acidic vesicular organelle cell autophagy marker, manifested by acridine orange (AO) and monodansylcadaverine (MDC) staining and cleavage of microtubule-associated protein 1 light chain 3 (LC3). The protein expression levels of the Atg 5, Atg 7, Atg 12, Beclin-1 and LC3-II were upregulated in T24 cells after baicalin treatment. Inhibition of autophagy by 3-methyl-adenine (an inhibitor of class III phosphatidylinositol-3 kinase; 3-MA) reduced the cleavage of LC3 in T24 cells after baicalin treatment. Furthermore, protein expression levels of phospho-AKT (Ser473) and enzyme activity of AKT were downregulated in T24 cells after baicalin treatment. In conclusion, baicalin triggered cell autophagy through the AKT signaling pathway in T24 cells.',\n",
       "    'mesh_terms': ['Acridine Orange',\n",
       "     'Adenine',\n",
       "     'Apoptosis Regulatory Proteins',\n",
       "     'Autophagy',\n",
       "     'Autophagy-Related Protein 12',\n",
       "     'Autophagy-Related Protein 5',\n",
       "     'Autophagy-Related Protein 7',\n",
       "     'Beclin-1',\n",
       "     'Cadaverine',\n",
       "     'Caspase 3',\n",
       "     'Cell Line, Tumor',\n",
       "     'Cell Survival',\n",
       "     'Enzyme Inhibitors',\n",
       "     'Flavonoids',\n",
       "     'Humans',\n",
       "     'Medicine, Chinese Traditional',\n",
       "     'Membrane Proteins',\n",
       "     'Microtubule-Associated Proteins',\n",
       "     'Phosphoinositide-3 Kinase Inhibitors',\n",
       "     'Proto-Oncogene Proteins c-akt',\n",
       "     'Signal Transduction',\n",
       "     'Small Ubiquitin-Related Modifier Proteins',\n",
       "     'Transglutaminases',\n",
       "     'Ubiquitin-Activating Enzymes',\n",
       "     'Urinary Bladder Neoplasms']},\n",
       "   {'pmid': '24140394',\n",
       "    'title': 'Physico-chemical characterization and the in vitro genotoxicity of medical implants metal alloy (TiAlV and CoCrMo) and polyethylene particles in human lymphocytes.',\n",
       "    'abstract': 'BACKGROUND: The main objective of the present study was to investigate chemical composition and possible cyto/genotoxic potential of several medical implant materials commonly used in total hip joint replacement. METHODS: Medical implant metal alloy (Ti6Al4V and CoCrMo) and high density polyethylene particles were analyzed by energy dispersive X-ray spectrometry while toxicological characterization was done on human lymphocytes using multi-biomarker approach. RESULTS: Energy dispersive X-ray spectrometry showed that none of the elements identified deviate from the chemical composition defined by appropriate ISO standard. Toxicological characterization showed that the tested materials were non-cyto/genotoxic as determined by the comet and cytokinesis-block micronucleus (CBMN) assay. Particle morphology was found (by using scanning electron and optical microscope) as flat, sharp-edged, irregularly shaped fiber-like grains with the mean particle size less than 10microm; this corresponds to the so-called \"submicron wear\". The very large surface area per wear volume enables high reactivity with surrounding media and cellular elements. CONCLUSIONS: Although orthopedic implants proved to be non-cyto/genotoxic, in tested concentration (10mug/ml) there is a constant need for monitoring of patients that have implanted artificial hips or other joints, to minimize the risks of any unwanted health effects. GENERAL SIGNIFICANCE: The fractal and multifractal analyses, performed in order to evaluate the degree of particle shape effect, showed that the fractal and multifractal terms are related to the \"remnant\" level of the particles\\' toxicity especially with the cell viability (trypan blue method) and total number of nucleoplasmic bridges and nuclear buds as CBMN assay parameters.',\n",
       "    'mesh_terms': ['Alloys',\n",
       "     'Biocompatible Materials',\n",
       "     'Cell Survival',\n",
       "     'Cells, Cultured',\n",
       "     'Chromium',\n",
       "     'Cobalt',\n",
       "     'Comet Assay',\n",
       "     'Cytokinesis',\n",
       "     'DNA Damage',\n",
       "     'Humans',\n",
       "     'Lymphocytes',\n",
       "     'Materials Testing',\n",
       "     'Micronucleus Tests',\n",
       "     'Molybdenum',\n",
       "     'Polyethylene',\n",
       "     'Prostheses and Implants',\n",
       "     'Spectrometry, X-Ray Emission',\n",
       "     'Titanium']},\n",
       "   {'pmid': '24139500',\n",
       "    'title': 'Dipyrithione induces cell-cycle arrest and apoptosis in four cancer cell lines in vitro and inhibits tumor growth in a mouse model.',\n",
       "    'abstract': \"BACKGROUND: Dipyrithione (PTS2) is widely used as a bactericide and fungicide. Here, we investigated whether PTS2 has broad-spectrum antitumor activity by studying its cytotoxicity and proapoptotic effects in four cancer cell lines. METHODS: We used MTT assays and trypan blue staining to test the viability of cancer cell lines. Hoechst 33258 and DAPI staining were used to observe cell apoptosis. Cell-cycle percentages were analyzed by flow cytometry. Apoptosis was assayed using caspase-3 and poly (ADP-ribose) polymerase (PARP) combined with Western blotting. Student's t-test was used for statistical analysis. RESULTS: PTS2 inhibited proliferation in four cancer cell lines in a dose-dependent manner. Treated cells showed shrinkage, irregular fragments, condensed and dispersed blue fluorescent particles compared with control cells. PTS2 induced cycle-arrest and death. Cleavage of caspase-9, caspase-3, and PARP were detected in PTS2-treated cells. Antitumor activity of PTS2 was more effective against widely used cancer drugs and its precursor. CONCLUSIONS: PTS2 appears to have novel cytotoxicity and potent broad-spectrum antitumor activity, which suggests its potential as the basis of an anticancer drug.\",\n",
       "    'mesh_terms': ['Animals',\n",
       "     'Antineoplastic Agents',\n",
       "     'Apoptosis',\n",
       "     'Cell Cycle Checkpoints',\n",
       "     'Cell Proliferation',\n",
       "     'Flow Cytometry',\n",
       "     'Humans',\n",
       "     'K562 Cells',\n",
       "     'Male',\n",
       "     'Mice',\n",
       "     'Mice, Inbred ICR',\n",
       "     'Molecular Structure',\n",
       "     'Pyridines',\n",
       "     'U937 Cells',\n",
       "     'Xenograft Model Antitumor Assays']},\n",
       "   {'pmid': '24190701',\n",
       "    'title': 'Bortezomib-induced apoptosis in cultured pancreatic cancer cells is associated with ceramide production.',\n",
       "    'abstract': \"PURPOSE: The proteasome inhibitor bortezomib (PS-341) has displayed significant efficiency against pancreatic cancer cells. However, the underlying mechanisms are not fully understood. Here, we tested if ceramide production was involved in the bortezomib's effect. METHODS: Two transformed pancreatic cancer cell lines (PANC-1 and Mia) and the primary pancreatic cancer cells were used. Cell death was analyzed by MTT viability assay and trypan blue staining. Cell apoptosis was analyzed by Histone DNA-ELISA assay and Annexin V FACS. Western blots were used to test signal protein changes. The cellular ceramide level after bortezomib treatment was also determined. RESULTS: In cultured pancreatic cancer cells, bortezomib increased cellular ceramide production to promote cell apoptosis. The ceramide de novo synthase inhibitor fumonisin B1 (F-B1) suppressed bortezomib-induced ceramide production and apoptosis, while exogenously added C6-ceramide facilitated bortezomib-induced pancreatic cancer cell death. Meanwhile, 1-phenyl-2-decanoylamino-3-morpholino-1-propanol (PDMP), the inhibitor of glucosylceramide synthetase as well as the sphingosine kinase 1 inhibitors (SKI-II and SKI-IV), facilitated bortezomib-induced ceramide production and subsequent cell apoptosis. Further, bortezomib-induced pro-apoptotic c-Jun N-terminal kinase (JNK) activation was also associated with ceramide production. JNK activation by bortezomib was suppressed by F-B1, but was enhanced by SKI-II and PDMP in pancreatic cancer cells. Finally, C6-ceramide, SKI-II, and PDMP dramatically enhanced bortezomib-induced cytotoxicity in primary cultured pancreatic cancer cells. CONCLUSIONS: We found that bortezomib-induced apoptosis was associated with ceramide production in primary and transformed pancreatic cancer cells.\",\n",
       "    'mesh_terms': ['Antineoplastic Agents',\n",
       "     'Apoptosis',\n",
       "     'Boronic Acids',\n",
       "     'Bortezomib',\n",
       "     'Cell Line, Tumor',\n",
       "     'Ceramides',\n",
       "     'Humans',\n",
       "     'JNK Mitogen-Activated Protein Kinases',\n",
       "     'MAP Kinase Kinase Kinase 5',\n",
       "     'Morpholines',\n",
       "     'Pancreatic Neoplasms',\n",
       "     'Phosphotransferases (Alcohol Group Acceptor)',\n",
       "     'Pyrazines']},\n",
       "   {'pmid': '24312318',\n",
       "    'title': 'mtDNA T8993G mutation-induced F1F0-ATP synthase defect augments mitochondrial dysfunction associated with hypoxia/reoxygenation: the protective role of melatonin.',\n",
       "    'abstract': 'BACKGROUND: F1F0-ATP synthase (F1F0-ATPase) plays important roles in regulating mitochondrial function during hypoxia, but the effect of F1F0-ATPase defect on hypoxia/reoxygenation (H/RO) is unknown. The aim of this study was to investigate how mtDNA T8993G mutation (NARP)-induced inhibition of F1F0-ATPase modulates the H/RO-induced mitochondrial dysfunction. In addition, the potential for melatonin, a potent antioxidant with multiple mitochondrial protective properties, to protect NARP cells exposed to H/RO was assessed. METHODS AND FINDINGS: NARP cybrids harboring 98% of mtDNA T8993G genes were established as an in vitro model for cells with F1F0-ATPase defect; their parental osteosarcoma 143B cells were studied for comparison. Treating the cells with H/RO using a hypoxic chamber resembles ischemia/reperfusion in vivo. NARP significantly enhanced apoptotic death upon H/RO detected by MTT assay and the trypan blue exclusion test of cell viability. Based on fluorescence probe-coupled laser scanning imaging microscopy, NARP significantly enhanced mitochondrial reactive oxygen species (mROS) formation and mitochondrial Ca(2+) (mCa(2+)) accumulation in response to H/RO, which augmented the depletion of cardiolipin, resulting in the retardation of mitochondrial movement. With stronger H/RO stress (either with longer reoxygenation duration, longer hypoxia duration, or administrating secondary oxidative stress following H/RO), NARP augmented H/RO-induced mROS formation to significantly depolarize mitochondrial membrane potential (DeltaPsim), and enhance mCa(2+) accumulation and nitric oxide formation. Also, NARP augmented H/RO-induced mROS oxidized and depleted cardiolipin, thereby promoting permanent mitochondrial permeability transition, retarded mitochondrial movement, and enhanced apoptosis. Melatonin markedly reduced NARP-augmented H/RO-induced mROS formation and therefore significantly reduced mROS-mediated depolarization of DeltaPsim and accumulation of mCa(2+), stabilized cardiolipin, and then improved mitochondrial movement and cell survival. CONCLUSION: NARP-induced inhibition of F1F0-ATPase enhances mROS formation upon H/RO, which augments the depletion of cardiolipin and retardation of mitochondrial movement. Melatonin may have the potential to rescue patients with ischemia/reperfusion insults, even those associated with NARP symptoms.',\n",
       "    'mesh_terms': ['Apoptosis',\n",
       "     'Calcium',\n",
       "     'Cardiolipins',\n",
       "     'Cell Hypoxia',\n",
       "     'Cell Line, Tumor',\n",
       "     'DNA, Mitochondrial',\n",
       "     'Humans',\n",
       "     'Melatonin',\n",
       "     'Mitochondria',\n",
       "     'Mitochondrial Proton-Translocating ATPases',\n",
       "     'Mutation',\n",
       "     'Neuroprotective Agents',\n",
       "     'Oxygen',\n",
       "     'Reactive Oxygen Species']}]}]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "issued_data = [data for data in graph_data if data['id'] == 1159]\n",
    "issued_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Initializations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-08 22:58:37,704 [DEBUG] embedding_model - CUDA is available, using GPU\n",
      "2025-04-08 22:58:58,096 [DEBUG] embedding_model - Embedding model initialized: neuml/pubmedbert-base-embeddings\n",
      "2025-04-08 22:58:58,106 [DEBUG] llm - Initialized model gemini-2.0-flash-lite\n",
      "2025-04-08 22:59:02,211 [DEBUG] connection - Connection successful!\n",
      "2025-04-08 22:59:02,257 [INFO] reader - Limiting the number of rows to 3...\n",
      "2025-04-08 22:59:02,259 [INFO] reader - Data file loaded with shape: (3, 4)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data length: 3\n"
     ]
    }
   ],
   "source": [
    "# models\n",
    "embedding_model = EmbeddingModel()\n",
    "\n",
    "llm = ChatModel(provider=\"google\", model_name=\"gemini-2.0-flash-lite\").initialize_model()\n",
    "\n",
    "# neo4j connection\n",
    "neo4j_connection = Neo4jConnection(uri=ConfigEnv.NEO4J_URI, \n",
    "                 user=ConfigEnv.NEO4J_USER,\n",
    "                 password=ConfigEnv.NEO4J_PASSWORD,\n",
    "                 database=ConfigEnv.NEO4J_DB)\n",
    "\n",
    "# retriever\n",
    "vector_search_tool = VectorSimilaritySearchTool(\n",
    "    llm=llm,\n",
    "    embedding_model=embedding_model,\n",
    "    neo4j_connection=neo4j_connection,\n",
    "    return_direct=True,\n",
    "    k=4\n",
    ")\n",
    "\n",
    "similarity_retriever = SimilaritySearchRetriever(\n",
    "    llm=llm,\n",
    "    embedding_model=embedding_model,\n",
    "    neo4j_connection=neo4j_connection,\n",
    ")\n",
    "\n",
    "retriever_model_name = vector_search_tool.get_model_name()\n",
    "\n",
    "# data\n",
    "data_path = os.path.join(ConfigPath.RAW_DATA_DIR, \"bioasq_train.parquet\")\n",
    "reader = BioASQDataReader(samples_limit=3)\n",
    "data = reader.read_parquet_file(file_path=data_path) \n",
    "print(f\"Data length: {len(data)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3529',\n",
       "  'pmid': '24348390',\n",
       "  'content': \"Glioblastoma multiforme (GBM) is the most lethal subtype of glioma, classified as a WHO grade 4 infiltrative glioma. The etiology of GBM remains unknown and risk factors can be identified only in a small minority. We report the synchronous occurrence of GBM in an otherwise unrelated married couple, i.e. a husband and his wife, who developed GBM within an interval of 1 month. No specific causative environmental factors were identified for both patients, and the genetic screens were negative for hereditary syndromes. Family history was negative for tumors, and no other incidence of cancer in either siblings, parents or other children was reported. An analysis of the couple's exposure to nonionizing electromagnetic fields and ionizing radiations revealed values within the normal ranges usually found in homes. Overall, conjugal tumors are rarely reported. However, the case reported herein raises important questions about possible etiologic factors.\",\n",
       "  'score': 0.7882180213928223},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3526',\n",
       "  'pmid': '22882019',\n",
       "  'content': 'A practical mathematical model for glioblastomas (brain tumours), which incorporates the two key parameters of tumour growth, namely the cancer cell diffusion and the cell proliferation rate, has been shown to be clinically useful and predictive. Previous studies explain why multifocal recurrence is inevitable and show how various treatment scenarios have been incorporated in the model. In most tumours, it is not known when the cancer started. Based on patient in vivo parameters, obtained from two brain scans, it is shown how to estimate the time, after initial detection, when the tumour started. This is an input of potential importance in any future controlled clinical study of any connection between cell phone radiation and brain tumour incidence. It is also used to estimate more accurately survival times from detection. Finally, based on patient parameters, the solution of the model equation of the tumour growth helps to explain why certain patients live longer than others after similar treatment protocols specifically surgical resection (removal) and irradiation.',\n",
       "  'score': 0.7499463558197021},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3536',\n",
       "  'pmid': '23261330',\n",
       "  'content': \" the Hardell group studies yielding in the >10 years latency period hazard ratio (HR)=1.2, 95% CI=1.002-1.5 for use of wireless phones. This increased HR was based on results for astrocytoma WHO grade IV (glioblastoma multiforme). Decreased HR was found for low-grade astrocytoma, WHO grades I-II, which might be caused by RF-EMF exposure leading to tumour-associated symptoms and earlier detection and surgery with better prognosis. Some studies show increasing incidence of brain tumours whereas other studies do not. It is concluded that one should be careful using incidence data to dismiss results in analytical epidemiology. The IARC carcinogenic classification does not seem to have had any significant impact on governments' perceptions of their responsibilities to protect public health from this widespread source of radiation.\",\n",
       "  'score': 0.6956987380981445},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3531',\n",
       "  'pmid': '23095687',\n",
       "  'content': 'BACKGROUND: We analysed the survival of patients after glioma diagnosis in relation to the use of wireless phones. METHODS: All cases diagnosed between 1997 and 2003 with a malignant brain tumour (n = 1,251) in our case-control studies were included and followed from the date of diagnosis to the date of death or until May 30, 2012. RESULTS: For glioma, the use of wireless phones (mobile and cordless phones) gave a hazard ratio (HR) = 1.1 (95% confidence interval, CI = 0.9-1.2), with > 10-year latency HR = 1.2 (95% CI = 1.002-1.5, p trend = 0.02). For astrocytoma grade I-II (low-grade), the results were, HR = 0.5 (95% CI = 0.3-0.9) and for astrocytoma grade IV (glioblastoma), HR = 1.1 (95% CI = 0.95-1.4), with > 10 year latency HR = 1.3 (95% CI = 1.03-1.7). In the highest tertile (> 426 h) of cumulative use, HR = 1.2 (95% CI = 0.95-1.5) was found for glioblastoma. The results were similar for mobile and cordless phones. CONCLUSIONS: Decreased survival of glioma cases with long-term and high cumulative use of wireless phones was found. A survival disadvantage for astrocytoma grade IV, but a survival benefit for astrocytoma grade I-II was observed which could be due to exposure-related tumour symptoms leading to earlier diagnosis and surgery in that patient group.',\n",
       "  'score': 0.6744163036346436}]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "similarity_retriever.get_relevant_contexts(query=\"Glioblastoma multiforme (GBM) is\", k=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = similarity_retriever.get_1_hop_similar_contexts(query=\"Circular RNAs (circRNAs) are a large type of noncoding RNAs \", k=10, n_similar_contexts=5)\n",
    "len(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3780',\n",
       "  'pmid': '28634583',\n",
       "  'content': \"Circular RNAs (circRNAs) are currently classed as non-coding RNA (ncRNA) that, unlike linear RNAs, form covalently closed continuous loops and act as gene regulators in mammals. They were originally thought to represent errors in splicing and considered to be of low abundance, however, there is now an increased appreciation of their important function in gene regulation. circRNAs are differentially generated by backsplicing of exons or from lariat introns. Unlike linear RNA, the 3' and 5' ends normally present in an RNA molecule have been joined together by covalent bonds leading to circularization. Interestingly, they have been found to be abundant, evolutionally conserved and relatively stable in the cytoplasm. These features confer numerous potential functions to circRNAs, such as acting as miRNA sponges, or binding to RNA-associated proteins to form RNA-protein complexes that regulate gene transcription. It has been proposed that circRNA regulate gene expression at the transcriptional or post-transcriptional level by interacting with miRNAs and that circRNAs may have a role in regulating miRNA function in cancer initiation and progression. circRNAs appear to be more often downregulated in tumor tissue compared to normal tissue and this may be due to (i) errors in the back-splice machinery in malignant tissues, (ii) degradation of circRNAs by deregulated miRNAs in tumor tissue, or (iii) increasing cell proliferation leading to a reduction of circRNAs. circRNAs have been identified in exosomes and more recently, chromosomal translocations in cancer have been shown to generate aberrant fusion-circRNAs associated with resistance to drug treatments. In addition, though originally thought to be non-coding, there is now increasing evidence to suggest that select circRNAs can be translated into functional proteins. Although much remains to be elucidated about circRNA biology and mechanisms of gene regulation, these ncRNAs are quickly emerging as potential disease biomarkers and therapeutic targets in cancer.\",\n",
       "  'score': 0.9184248447418213},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3651',\n",
       "  'pmid': '27612318',\n",
       "  'content': 'Circular RNAs (circRNAs) represent a large class of noncoding RNAs (ncRNAs) that have recently emerged as regulators of gene expression. They have been shown to suppress microRNAs, thereby increasing the translation and stability of the targets of such microRNAs. In this review, we discuss the emerging functions of circRNAs, including RNA transcription, splicing, turnover, and translation. We also discuss other possible facets of circRNAs that can influence their function depending on the cell context, such as circRNA abundance, subcellular localization, interacting partners (RNA, DNA, and proteins), dynamic changes in interactions following stimulation, and potential circRNA translation. The ensuing changes in gene expression patterns elicited by circRNAs are proposed to drive key cellular processes, such as cell proliferation, differentiation, and survival, that govern health and disease. WIREs RNA 2017, 8:e1386. doi: 10.1002/wrna.1386 For further resources related to this article, please visit the WIREs website.',\n",
       "  'score': 0.9163014888763428},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3649',\n",
       "  'pmid': '27617908',\n",
       "  'content': 'Circular RNAs (circRNAs) are a novel class of non-coding RNA that assumes a covalently closed continuous conformation. CircRNAs were previously thought to be the byproducts of splicing errors caused by low abundance and the technological limitations. With the recent development of high-throughput sequencing technology, numerous circRNAs have been discovered in many species. Recent studies have revealed that circRNAs are stable and widely expressed, and often exhibit cell type-specific or tissue-specific expression. Most circRNAs can be generated from exons, introns, or both. Remarkably, emerging evidence indicates that some circRNAs can serve as microRNA (miRNA) sponges, regulate transcription or splicing, and can interact with RNA binding proteins (RBPs). Moreover, circRNAs have been reported to play essential roles in myriad life processes, such as aging, insulin secretion, tissue development, atherosclerotic vascular disease risk, cardiac hypertrophy and cancer. Although circRNAs are ancient molecules, they represent a newly appreciated form of noncoding RNA and as such have great potential implications in clinical and research fields. Here, we review the current understanding of circRNA classification, function and significance in physiological and pathological processes. We believe that future research will increase our understanding of the regulation and function of these novel molecules.',\n",
       "  'score': 0.9138593673706055},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3644',\n",
       "  'pmid': '27255916',\n",
       "  'content': 'Insights into the circular RNA (circRNA) exploration have revealed that they are abundant in eukaryotic transcriptomes. Diverse genomic regions can generate different types of RNA circles, implying their diversity. Covalently closed loop structures elevate the stability of this new type of noncoding RNA. High-throughput sequencing analyses suggest that circRNAs exhibit tissue- and developmental-specific expression, indicating that they may play crucial roles in multiple cellular processes. Strikingly, several circRNAs could function as microRNA sponges and regulate gene transcription, highlighting a new class of important regulators. Here, we review the recent advances in knowledge of endogenous circRNA biogenesis, properties and functions. We further discuss the current findings about circRNAs in human diseases. In plants, the roles of circRNAs remain a mystery. Online resources and bioinformatics identification of circRNAs are essential for the analysis of circRNA biology, although different strategies yield divergent results. The understanding of circRNA functions remains limited; however, circRNAs are enriching the RNA world, acting as an emerging key player.',\n",
       "  'score': 0.9138023853302002},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3767',\n",
       "  'pmid': '30550956',\n",
       "  'content': 'Circular RNAs (circRNAs) are a novel class of regulatory RNAs that despite being relatively abundant have only recently begun to be explored. There are many thousands of genes that appear capable of producing circRNAs, however the function of all but a handful remain to be determined. What is emerging about these highly conserved molecules is that they play important roles in biology and cancer biology in particular. The most explored function of circRNAs is as master regulators of gene expression that act to sequester or  sponge  other gene expression regulators, in particular miRNAs. They have also been demonstrated to function via direct modulation of transcription, and by interfering with splicing mechanisms. Although generally expressed in low abundance when compared to their linear counterparts, they are often expressed in a tissue- and developmental stage- specific manner. Coupled with their remarkable resistance to RNAse activity due to a covalent closed cyclic structure, circRNAs show great promise as novel biomarkers of cancer and other diseases. In this review we consider the current state of knowledge regarding these molecules, their synthesis, function, and association with cancer. We will also review some of the challenges that remain to be resolved if this emerging class of RNAs are really to become useful in the clinic.',\n",
       "  'score': 0.9075369834899902},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3760',\n",
       "  'pmid': '28969093',\n",
       "  'content': \"Circular RNAs (circRNAs) are a novel type of universal and diverse endogenous noncoding RNAs (ncRNAs) and they form a covalently closed continuous loop without 5' or 3' tails unlike linear RNAs. Most circRNAs are presented with characteristics of abundance, stability, conservatism, and often exhibiting tissue/developmental-stage-specific expression. CircRNAs are generated either from exons or introns by back splicing or lariat introns. CircRNAs play important roles as miRNA sponges, gene transcription and expression regulators, RNA-binding protein (RBP) sponges and protein/peptide translators. Emerging evidence revealed the function of circRNAs in cancer and may potentially serve as a required novel biomarker and therapeutic target for cancer treatment. In this review, we discuss about the origins, characteristics and functions of circRNA and how they work as miRNA sponges, gene transcription and expression regulators, RBP sponges in cancer as well as current research methods of circRNAs, providing evidence for the significance of circRNAs in cancer diagnosis and clinical treatment.\",\n",
       "  'score': 0.9064779281616211},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3782',\n",
       "  'pmid': '28018143',\n",
       "  'content': 'Many thousands of Circular RNAs (circRNAs) have recently been identified in metazoan genomes by transcriptome-wide sequencing. Most circRNAs are generated by back-splicing events from exons of protein-coding genes. A great deal of progress has recently been made in understanding the genome-wide expression patterns, biogenesis, and regulation of circRNAs. To date, however, few functions of circRNAs have been identified. CircRNAs are preferentially expressed in neural tissues and some are found at synapses, suggesting possible functions in the nervous system. Several circRNAs have been shown to function as microRNA \"sponges\" to counteract microRNA mediated repression of mRNA. New functions for circRNAs are arising, including protein sequestration, transcriptional regulation, and potential functions in cancer. Here, we highlight the recent progress made in understanding the biogenesis and regulation of circRNAs, discuss newly uncovered circRNA functions, and explain the methodological approaches that could reveal more exciting and unexpected roles for these RNAs.',\n",
       "  'score': 0.8957300186157227},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3647',\n",
       "  'pmid': '26874353',\n",
       "  'content': 'Circular RNAs (circRNAs) are a large type of noncoding RNAs characterized by their circular shape resulting from covalently closed continuous loops. They are known to regulate gene expression in mammals. These tissue-specific transcripts are largely generated from exonic or intronic sequences of their host genes. Although several models of circRNA biogenesis have been proposed, the understanding of their origin is far from complete. Unlike other noncoding RNAs, circRNAs are widely expressed, highly conserved and stable in cytoplasm, which confer special functionalities to them. They are known to serve as microRNA (miRNA) sponges, regulators of alternative splicing, transcription factors and encode for proteins. The expression of circRNAs is associated with several pathological states and may potentially serve as novel diagnostic or predictive biomarkers. CircRNAs are known to regulate the expression of numerous cancer-related miRNAs. The circRNA-miRNA-mRNA axis is a known regulatory pattern of several cancer-associated pathways, with both agonist and antagonist effects on carcinogenesis. In consideration of their potential clinical relevance, circRNAs are at the center of ongoing research initiatives on cancer prevention and treatment. In this review, we discuss the current understanding of circRNAs and the prospects for their potential clinical application in the management of cancer patients.',\n",
       "  'score': 0.8956494331359863},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3775',\n",
       "  'pmid': '29349062',\n",
       "  'content': \"Circular RNAs (circRNAs) are a class of noncoding RNAs (ncRNAs) that form covalently closed continuous loop structures, lacking the terminal 5' and 3' ends. CircRNAs are generated in the process of back-splicing and can originate from different genomic regions. Their unique circular structure makes circRNAs more stable than linear RNAs. In addition, they also display insensitivity to ribonuclease activity. Generally, circRNAs function as microRNA (miRNA) sponges and have a regulatory role in transcription and translation. They may be also translated in a cap-independent manner in vivo, to generate specific proteins. In the last decade, next-generation sequencing techniques, especially RNA-seq, have revealed great abundance and also dysregulation of many circRNAs in various diseases, suggesting their involvement in disease development and progression. Regarding their high stability and relatively specific differential expression patterns in tissues and extracellular environment (e.g., body fluids), they are regarded as promising novel biomarkers in cancer. Therefore, we focus this review on describing circRNA biogenesis, function, and involvement in human cancer development and address the potential of circRNAs to be effectively used as novel cancer diagnostic and prognostic biomarkers.\",\n",
       "  'score': 0.8914563655853271},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3645',\n",
       "  'pmid': '26649774',\n",
       "  'content': 'Circular RNAs (circRNAs) are long, non-coding RNAs that result from the non-canonical splicing of linear pre-mRNAs. However, the characteristics and the critical role of circRNA in co-/post-transcriptional regulation were not well recognized until the \"microRNA sponge\" function of circRNA is discovered. Recent studies have mainly been devoted to the function of the circular RNA sponge for miR-7 (ciRS-7) and sex-determining region Y (SRY) by targeting microRNA-7 (miR-7) and microRNA-138 (miR-138), respectively. In this review, we illustrate the specific role of circRNAs in a wide variety of cancers and in regulating the biological behavior of cancers via miR-7 or miR-138 regulation. Furthermore, circRNA, together with its gene silencing ability, also shows its potential in RNA interference (RNAi) therapy by binding to target RNAs, which provides a novel perspective in cancer treatment. Thus, this review concerns the biogenesis, biological function, oncogenesis, progression and possible therapies for cancer involving circRNAs.',\n",
       "  'score': 0.8887155055999756},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3776',\n",
       "  'pmid': '29387208',\n",
       "  'content': 'Circular (circ)RNAs, naturally formed endogenous non-coding RNAs, have received extensive attention in recent years due to their special loop structures and specific function. circRNAs are formed with covalently closed continuous loops and are mainly generated by back-splicing processes or lariat introns from exons and/or introns. Usually, circRNAs are stable, abundant, and evolutionarily conserved in the cytoplasm. circRNAs often exhibit abnormal expression in different diseases, notably in human cancers, and the presence of abundant circRNAs in serum, saliva and exosomes renders them potential diagnostic or predictive biomarkers for diseases, including multiple types of cancer. Presently, certain circRNAs have been reported to function as microRNA sponges and RNA-binding protein sponges to regulate downstream gene transcription, which suggests a potential for circRNAs in cancer diagnosis, prognosis and clinical therapy. The present study assessed the latest advances in the study of circRNAs in cancer, summarized the functions of circRNAs in different types of cancer, highlighted the competing endogenous RNA function of circRNAs in the occurrence and development of human malignancies, and provided evidence for the future application of circRNAs in the diagnosis, prognosis and treatment of multiple types of cancer.',\n",
       "  'score': 0.940154380485886},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3779',\n",
       "  'pmid': '33537235',\n",
       "  'content': 'Circular RNAs (circRNAs) are a class of endogenous single-stranded covalently closed RNAs, primarily produced from pre-mRNAs via non-canonical back-splicing. circRNAs are highly conserved, stable, and expressed in tissue- and development-specific pattern. circRNAs play essential roles in physiological process as well as cancer biology. By the advances of deep sequencing and bioinformatics, the number of circRNAs have increased explosively. circRNAs function as miRNA/protein sponge, protein scaffold, protein recruitment, enhancer of protein function, as well as templates for translation involved in the regulation of transcription/splicing, translation, protein degradation, and pri-miRNA processing in human cancers and contributed to the pathogenesis of cancer. Numerous circRNAs may function in diverse manners. In this review, we survey the current understanding of circRNA functions in human cancer including miRNA sponge, circRNA-protein interaction, and circRNA-encoded protein, and summarize available databases for circRNA annotation and functional prediction.',\n",
       "  'score': 0.9235913475882136},\n",
       " {'element_id': '4:d362a474-c8ba-449f-aeb1-9e0df44dbd24:3778',\n",
       "  'pmid': '27616979',\n",
       "  'content': 'Circular RNAs (circRNAs) are highly stable, circularized long non-coding RNAs. circRNAs are conserved across species and appear to be specifically enriched in the nervous system. Recent studies show that many circRNAs are expressed in a tissue- and developmental-stage-specific manner, reveal a striking regulation of circRNAs during neuronal development, and detect their presence at synaptic sites. The exact functions of circRNAs remain poorly understood, but evidence from analysis of some circRNA molecules suggests that they could substantially contribute to the regulation of gene expression, particularly in architecturally complex and polarized cells such as neurons. Emerging evidence also indicates that circRNAs are involved in the development and progression of various neurological disorders. In this review, we summarize the molecular characteristics of circRNAs and discuss their proposed functions and mechanism-of-action in developing neurons.',\n",
       "  'score': 0.9081021962027914}]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'a': 2}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def func1(**kwargs):\n",
    "    print(kwargs)\n",
    "  \n",
    "func_args = {\"a\":2}  \n",
    "func1(**func_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "existing_index = Neo4jVector.from_existing_index(\n",
    "    embedding_model,\n",
    "    url=ConfigEnv.NEO4J_URI,\n",
    "    username=ConfigEnv.NEO4J_USER,\n",
    "    password=ConfigEnv.NEO4J_PASSWORD,\n",
    "    index_name=\"contextIndex\",\n",
    "    text_node_property=\"text_content\",  # Need to define if it is not default\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "existing_index.similarity_search(\"hey\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "neo4j_connection.execute_query(\"\"\"SHOW INDEXES\n",
    "       YIELD name, type, labelsOrTypes, properties, options\n",
    "       WHERE type = 'VECTOR'\"\"\", params={})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "index_name = \"context_index\"  # default index name\n",
    "\n",
    "existing_graph = Neo4jVector.from_existing_graph(\n",
    "    embedding=embedding_model,\n",
    "    url=ConfigEnv.NEO4J_URI,\n",
    "    username=ConfigEnv.NEO4J_USER,\n",
    "    password=ConfigEnv.NEO4J_PASSWORD,\n",
    "    index_name=\"context_index\",\n",
    "    node_label=\"CONTEXT\",\n",
    "    # text_node_properties=[\"text_content\"],\n",
    "    embedding_node_property=\"embedding\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "existing_graph.search(\"Utilization behavior (UB) consists of reaching out and using objects in the environment in an automatic mann\", search_type=\"mmr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def run_retriever(benchmark_data: list, retriever) -> dict:\n",
    "    results = {}\n",
    "    for sample in tqdm(benchmark_data, desc=\"Executing retriever...\"):\n",
    "        sample_id = sample.get('id')\n",
    "        question = sample.get('question')\n",
    "        retrieved_data = retriever.invoke(question)\n",
    "        results[sample_id] = retrieved_data\n",
    "    return results\n",
    "\n",
    "results = run_retriever(benchmark_data=data, retriever=vector_search_tool)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "metrics, new_evaluator = run_evaluation_on_retrieved_chunks(\n",
    "    retrieval_results=results,\n",
    "    benchmark_data=data\n",
    ")\n",
    "print(metrics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RAGAS METRICS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vector_search_tool = VectorSimilaritySearchTool(\n",
    "        llm=llm,\n",
    "        embedding_model=embedding_model,\n",
    "        neo4j_connection=neo4j_connection,\n",
    "        return_direct=False,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = vector_search_tool.invoke(\"What is the implication of histone lysine methylation in medulloblastoma?\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(results['answer'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "contexts = [sample['content'] for sample in results['context']]\n",
    "contexts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ragas.llms import LangchainLLMWrapper\n",
    "\n",
    "llm = ChatModel(provider=\"google\", model_name=\"gemini-2.0-flash-lite\").initialize_model()\n",
    "evaluator_llm = LangchainLLMWrapper(llm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from ragas.dataset_schema import SingleTurnSample\n",
    "from ragas.metrics import ContextRelevance\n",
    "\n",
    "sample = SingleTurnSample(\n",
    "    user_input=\"What is the implication of histone lysine methylation in medulloblastoma?\",\n",
    "    retrieved_contexts=contexts\n",
    ")\n",
    "\n",
    "scorer = ContextRelevance(llm=evaluator_llm)\n",
    "score = await scorer.single_turn_ascore(sample)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ragas import EvaluationDataset\n",
    "\n",
    "\n",
    "dataset = []\n",
    "dataset.append(\n",
    "        {\n",
    "            \"id\": 1,\n",
    "            \"user_input\": \"What is the implication of histone lysine methylation in medulloblastoma?\",\n",
    "            \"retrieved_contexts\": contexts,\n",
    "            \"response\": \"Histone lysine methylation, particularly at H3K9, is implicated in the pathogenesis of medulloblastoma. Copy number aberrations in genes involved in writing, reading, removing, and blocking histone lysine methylation suggest that defective control of the histone code contributes to the development of this cancer. Additionally, the study found that restoration of expression of genes controlling H3K9 methylation greatly diminishes proliferation of medulloblastoma in vitro.\",\n",
    "            \"reference\": \"Aberrant patterns of H3K4, H3K9, and H3K27 histone lysine methylation were shown to result in histone code alterations, which induce changes in gene expression, and affect the proliferation rate of cells in medulloblastoma.\", # expected response\n",
    "        }\n",
    "    )\n",
    "evaluation_dataset = EvaluationDataset.from_list(dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ragas import evaluate\n",
    "from ragas.llms import LangchainLLMWrapper\n",
    "from ragas.embeddings import LangchainEmbeddingsWrapper\n",
    "from ragas.metrics import LLMContextPrecisionWithReference, ContextRecall, ResponseRelevancy, FactualCorrectness\n",
    "\n",
    "evaluator_llm = LangchainLLMWrapper(llm)\n",
    "evaluator_embedding = LangchainEmbeddingsWrapper(embedding_model)\n",
    "\n",
    "context_precision = LLMContextPrecisionWithReference()\n",
    "context_recall = ContextRecall()\n",
    "response_relevancy = ResponseRelevancy()\n",
    "factual_correctness = FactualCorrectness()\n",
    "\n",
    "result = evaluate(\n",
    "    dataset=evaluation_dataset,\n",
    "    metrics=[context_precision, context_recall, response_relevancy, factual_correctness],\n",
    "    llm=evaluator_llm,\n",
    "    embeddings=evaluator_embedding\n",
    ")\n",
    "\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read paquet data\n",
    "import os\n",
    "import pandas as pd\n",
    "from configs.config import ConfigPath\n",
    "\n",
    "from data_collection.reader import BioASQDataReader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "asq_reader = BioASQDataReader()\n",
    "data = asq_reader.read_parquet_file(file_path=os.path.join(ConfigPath.RAW_DATA_DIR, \"bioasq_train.parquet\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample in data:\n",
    "    if 20007090 in sample[\"relevant_passage_ids\"]:\n",
    "        print(sample)\n",
    "        print(data.index(sample))\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "asq_reader.get_data_to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from data_collection.fetcher import PubMedArticleFetcher\n",
    "\n",
    "fetcher = PubMedArticleFetcher()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = fetcher.fetch_articles(pmids=['20007090'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
